solidity/ecdsa/test/WalletRegistry.RandomBeacon.test.ts

```diff
@@ -0,0 +1,239 @@
+        await expect(randomBeaconFake.requestRelayEntry).to.be.calledWith(
+          walletRegistry.address
+        )
+      })
+    })
+  })
+
+  describe("__beaconCallback", async () => {
+    before(async () => {
+      await createSnapshot()
+    })
+
+    after(async () => {
+      await restoreSnapshot()
+    })
+
+    context("when called by a third party", async () => {
+      it("should revert", async () => {
+        await expect(
+          walletRegistry.connect(thirdParty).__beaconCallback(123, 456)
+        ).to.be.revertedWith("Caller is not the Random Beacon")
+      })
+    })
+
+    context("when called by the random beacon", async () => {
+      context("when new wallet was not requested", async () => {
+        it("should revert", async () => {
+          await expect(
+            walletRegistry
+              .connect(randomBeaconFake.wallet)
+              .__beaconCallback(123, 456)
+          ).to.be.revertedWith("Current state is not AWAITING_SEED")
+        })
+      })
+
+      context("when new wallet was requested", async () => {
+        const relayEntry = ethers.BigNumber.from(ethers.utils.randomBytes(32))
+        let tx: ContractTransaction
+
+        before(async () => {
+          await createSnapshot()
+
+          await walletRegistry.connect(walletOwner).requestNewWallet()
+
+          tx = await walletRegistry
+            .connect(randomBeaconFake.wallet)
+            .__beaconCallback(relayEntry, 0)
+        })
+
+        after(async () => {
+          await restoreSnapshot()
+        })
+
+        it("should transition wallet creation state to `AWAITING_RESULT`", async () => {
+          await expect(
+            await walletRegistry.getWalletCreationState()
+          ).to.be.equal(dkgState.AWAITING_RESULT)
+        })
+
+        it("should set seed for wallet creation", async () => {
+          await expect((await walletRegistry.getDkgData()).seed).to.be.equal(
+            relayEntry
+          )
+        })
+
+        it("should set start block for wallet creation", async () => {
+          await expect(
+            (
+              await walletRegistry.getDkgData()
+            ).startBlock
+          ).to.be.equal(tx.blockNumber)
+        })
+
+        it("should not emit DkgStateLocked event", async () => {
+          await expect(tx).not.to.emit(walletRegistry, "DkgStateLocked")
+        })
+
+        it("should emit DkgStarted event", async () => {
+          await expect(tx).to.emit(walletRegistry, "DkgStarted")
+        })
+      })
+
+      describe("gas estimation", async () => {
+        before(async () => {
+          await createSnapshot()
+
+          await walletRegistry.connect(walletOwner).requestNewWallet()
+        })
+
+        after(async () => {
+          await restoreSnapshot()
+        })
+
+        // The exact value was noted from a test execution and is used as a reference
+        // for all future executions.
+        it("should not exceed 77308", async () => {
```

Comment:
We got two types of tests in this file. Here we call the `__calback` function directly in a transaction, so it costs more than 56k as it adds ~20k as the transaction execution cost. Another way is like it will be used in real-life so executed inside another transaction so calling just the `__calback` function inside another function call will cost ~56k.